.eslintrc.cjs

```diff
@@ -64,12 +68,51 @@ module.exports = {
             allowNamedExports: true,
           },
         ],
-        "@typescript-eslint/no-unnecessary-type-assertion": "error",
         "@typescript-eslint/no-confusing-void-expression": [
           "error",
           { ignoreArrowShorthand: true },
         ],
         "@typescript-eslint/no-import-type-side-effects": "error",
+
+        // Todo: Investigate, for each of these, whether we want them
+        "@typescript-eslint/array-type": "off",
+        "@typescript-eslint/await-thenable": "off",
+        "@typescript-eslint/ban-types": "off",
+        "@typescript-eslint/consistent-generic-constructors": "off",
+        "@typescript-eslint/consistent-indexed-object-style": "off",
+        "@typescript-eslint/consistent-type-definitions": "off",
+        "@typescript-eslint/dot-notation": "off",
+        "@typescript-eslint/no-base-to-string": "off",
+        "@typescript-eslint/no-duplicate-type-constituents": "off",
+        "@typescript-eslint/no-empty-function": "off",
+        "@typescript-eslint/no-empty-interface": "off",
+        "@typescript-eslint/no-misused-promises": "off",
+        "@typescript-eslint/no-namespace": "off",
+        "@typescript-eslint/no-redundant-type-constituents": "off",
+        "@typescript-eslint/no-this-alias": "off",
+        "@typescript-eslint/no-unnecessary-type-assertion": "off",
```

Comment:
```suggestion
```
Glad that it is now enabled by default. Well, in that case...

Reply:
Yay! 

I tried this out a bit locally but ran into the following conundrum:

* Without `strictNullChecks: true` in `tsconfig.base.json`, fixing for `no-unnecessary-type-assertion` can cause issues.
    * Example: The `null as null` cast in `packages/babel-preset-env/src/shipped-proposals.ts` avoid a `Object literal's property '"transform-unicode-property-regex"' implicitly has an 'any' type.` complaint
* With `strictNullChecks: true` in `tsconfig.base.json`, there are quite a few complaints.
    * Example: `parentPath` in `packages/babel-traverse/src/path/modification.ts` > `insertBefore` is potentially `null` and that causes type errors

Can I propose filing a followup to enable `strictNullChecks: true` internally, and considering it a blocker for enabling some of these rules?

Reply:
Perhaps we can temporarily use `eslint-disable-next-line` with `strictNullChecks: false`?
I remember @JLHwung has a `strictNullChecks: true` branch, but it was blocked by ts bug at the time, and it may be outdated now.(I'm willing to give this a try too, if that branch is outdated)
By the way, can `null as null` be a special case of `no-unnecessary-type-assertion`?

Reply:
> temporarily use `eslint-disable-next-line` with `strictNullChecks: false`

That could work. In my experience that can bring positives (more lint rules, sooner) and also some negatives (folks getting confused when new edge case situations like this one pop up). What do you think?

> can `null as null` be a special case of `no-unnecessary-type-assertion`?

Interesting. We haven't had a lot of users running without `strictNullChecks`, and haven't had to think about this case before (that I know of). Similar with `undefined as undefined`...

I poked a bit around it and I think it's because I'd enabled `strict: true` in `tsconfig.eslint.json`. That made typescript-eslint run with different type information than TypeScript proper. We in typescript-eslint generally don't accept bugs or feature requests for the mismatched-type-information situation because it's so rare for users to hit it.

If I turn off `strictNullChecks` in `tsconfig.eslint.json` then we end up with a bunch of `no-unnecessary-type-assertion` complaints for code like `id as NodePath<t.Identifier | null>`. Which, while technically correct, seems like a backwards step towards your goal in enabling `strictNullChecks`.

Reply:
Makes sense! I'm slightly leaning towards enabling this rule and ignoring edge cases, but both options are fine.

It seems this is not just happening with this PR. 🤔
https://github.com/babel/babel/pull/15681/files?diff=unified#diff-2584b855363a673ff1758cc21653fdbd86b1ad2ab7f1a851c2320999c9c6f84cR30
I had the same problem when I upgraded TS, so maybe it's not a `tsconfig.eslint.json` specific issue.

Reply:
> Perhaps we can temporarily use `eslint-disable-next-line` with `strictNullChecks: false`? I remember @JLHwung has a `strictNullChecks: true` branch, but it was blocked by ts bug at the time, and it may be outdated now.(I'm willing to give this a try too, if that branch is outdated) By the way, can `null as null` be a special case of `no-unnecessary-type-assertion`?

Thanks for the reminder. I have rebased that branch but I still hits OOM: https://github.com/JLHwung/babel/tree/strict-null-checks, using TS 5.2.0 beta.

Reply:
![image](https://github.com/babel/babel/assets/30521560/4d6590ba-ceec-4cd0-9e4b-037b840821be)
This branch is using `TS 4.8 beta`. :)

Reply:
Good catch. Updated to 5.1.6 and it works for me now.